File: client/java/src/avt/CustomTab.java
package avt;

import java.util.Hashtable;
import java.util.Vector;
import javax.microedition.lcdui.Graphics;
import javax.microedition.lcdui.Image;
import main.Canvas;

public final class CustomTab extends Face {
   public static CustomTab me;
   private int x;
   private int xChange;
   private int y;
   private int w;
   private int h;
   private Vector listLabel = new Vector();
   private Vector listPic = new Vector();
   private Hashtable listImg;
   private String title = "";
   private String strTemp = "";
   private int x0 = 5;
   private int y0 = 5;
   private int wTab = 30;
   private byte idAction;
   private static int cmtoY;
   private static int cmy;
   private static int cmdy;
   private static int cmvy;
   private static int cmyLim;
   private static int yL;
   private static int wStr = 14;
   private long time;
   private int ww = 0;
   private int pa = 0;
   private boolean trans = false;
   private int vY;
   private int count;
   private int timePointY;
   private int dyTran;

   public static CustomTab gI() {
      return me == null ? (me = new CustomTab()) : me;
   }

   public final void commandTab(int var1, int var2) {
      switch (var1) {
         case 0:
            this.listLabel.removeAllElements();
            this.listPic.removeAllElements();
            this.listImg.clear();
            Canvas.currentFace = null;
            Canvas.endDlg();
            me = null;
            return;
         case 1:
            GlobalService.gI().doCustomTab(this.idAction);
         default:
      }
   }

   public CustomTab() {
      this.setSize();
      super.right = new Command(T.close, 0);
      wStr = AvMain.hBlack;
   }

   private void setSize() {
      this.w = Canvas.w - 20;
      this.h = Canvas.h - Canvas.hTab - 20;
   }

   public final void d() {
      this.setSize();
      this.setInfo(this.listImg, this.title, this.strTemp, this.idAction);
   }

   private void setFont(String var1) {
      int var2;
      if ((var2 = var1.indexOf("Ę")) != -1) {
         String var3 = var1.substring(0, var2);
         this.setCanhle(var3, "");
         if ((var2 = (var1 = var1.substring(var2 + 1)).indexOf("\n")) != -1) {
            var3 = var1.substring(0, var2);
            this.setCanhle(var3, "Ę");
            var1 = var1.substring(var2 + 1);
            this.setFont(var1);
         } else {
            this.setCanhle(var1, "Ę");
         }
      } else {
         this.setCanhle(var1, "");
      }

   }

   private void setCanhle(String var1, String var2) {
      if (!var1.equals("")) {
         int var3;
         if ((var3 = var1.indexOf("ę")) != -1) {
            String var4;
            if (!(var4 = var1.substring(0, var3)).equals("")) {
               this.addd(var4, var2, 0);
            }

            int var5 = Integer.parseInt(var1.substring(var3 + 1, var3 + 2));
            if ((var3 = (var1 = var1.substring(var3 + 2, var1.length())).indexOf("\n")) != -1) {
               this.addd(var1.substring(0, var3), var2, var5);
               this.setCanhle(var1.substring(var3 + 1), var2);
            } else {
               this.addd(var1, var2, var5);
            }
         } else {
            this.addd(var1, var2, 0);
         }
      }

   }

   private void addd(String var1, String var2, int var3) {
      boolean var4 = false;
      String[] var5;
      if (var1.indexOf("tem") != -1) {
         var4 = true;
         var5 = Canvas.fontChatB.splitFontBStrInLine(var1, this.w - 30 - 8 * AvMain.hd);
      } else {
         var5 = Canvas.normalFont.splitFontBStrInLine(var1, this.w - 30 - 8 * AvMain.hd);
      }

      for(int var6 = 0; var6 < var5.length; ++var6) {
         int var7;
         if (var4) {
            var7 = Canvas.fontChatB.getWidth(var5[var6]);
         } else {
            var7 = Canvas.normalFont.getWidth(var5[var6]);
         }

         if (var7 > this.ww) {
            this.ww = var7;
         }

         StringObj var8;
         (var8 = new StringObj(this.x0, this.y0 += wStr, var2 + var5[var6])).anthor = var3;
         this.listLabel.addElement(var8);
      }

   }

   public final void setInfo(Hashtable var1, String var2, String var3, byte var4) {
      this.count = 0;
      super.center = null;
      if (var4 != -1) {
         super.center = new Command(T.selectt, 1);
      }

      this.idAction = var4;
      this.ww = 0;
      this.wTab = Canvas.normalFont.getWidth(var2) + 20 * AvMain.hd;
      if (this.wTab < 50 + 20 * AvMain.hd) {
         this.wTab = 50 + 20 * AvMain.hd;
      }

      this.listImg = var1;
      this.title = var2;
      this.strTemp = var3;
      this.listLabel.removeAllElements();
      this.listPic.removeAllElements();
      boolean var9 = false;
      this.x0 = 0;
      this.y0 = -10;

      int var10;
      while((var10 = var3.indexOf("µ")) != -1) {
         String var13 = var3.substring(0, var10);
         var3 = var3.substring(var10 + 1, var3.length());
         String var5;
         if (var9) {
            var10 = var13.indexOf(",");
            var5 = var13.substring(0, var10);
            var10 = (var13 = var13.substring(var10 + 1, var13.length())).indexOf(",");
            int var6 = Integer.parseInt(var13.substring(0, var10));
            boolean var11 = Integer.parseInt(var13.substring(var10 + 1, var13.length())) != 0;
            Image var14 = (Image)this.listImg.get(var5);
            byte var7 = 0;
            if (var6 == 17) {
               var7 = 1;
            } else if (var6 == 24) {
               var7 = 2;
            }

            PictureObj var15 = new PictureObj(this, Integer.parseInt(var5), var7, this.y0 + wStr + 5, var6, var11);
            var14.getWidth();
            var15.e = var14.getHeight();
            if (var11) {
               PictureObj var12 = (PictureObj)this.listPic.elementAt(this.listPic.size() - 1);
               var6 = ((Image)this.listImg.get(String.valueOf(var12.d))).getHeight();
               if (var14.getHeight() > var6) {
                  var12.b += var14.getHeight() - var6;
               }

               var15.b = var12.b + var6 - var14.getHeight();
            }

            this.y0 = var15.b + var14.getHeight() - 10;
            this.listPic.addElement(var15);
            var13 = "";
         }

         var9 = !var9;

         while(true) {
            while((var10 = var13.indexOf("¶")) != -1) {
               var5 = var13.substring(0, var10);
               var13 = var13.substring(var10 + 1, var13.length());

               try {
                  Integer.parseInt(var5, 16);
                  this.setFont("¶" + var5);
                  this.y0 -= wStr / 2;
               } catch (Exception var15) {
                  this.setFont(var5);
               }
            }

            if (!var13.equals("")) {
               this.setFont(var13.substring(0, var13.length() - 1));
            }

            if (var3.indexOf("µ") != -1 || var3.indexOf("¶") == -1) {
               break;
            }

            var13 = var3;
            var3 = "";
         }
      }

      this.setFont(var3);
      this.xChange = 9 * AvMain.hd;
      if (this.ww < 140 * AvMain.hd) {
         this.ww = 140 * AvMain.hd;
      }

      if (this.ww >= 120 && this.ww < this.w - 30) {
         this.w = this.ww + 20 * AvMain.hd;
         this.xChange = 10 * AvMain.hd;
      }

      if (this.y0 + 10 + (wStr << 1) < this.h - 30) {
         this.h = this.y0 + 10 + (wStr << 1) + 20;
      }

      if (this.h < 80 * AvMain.hd + MyScreen.hTab) {
         this.h = 80 * AvMain.hd + MyScreen.hTab;
      }

      if ((cmyLim = this.y0 - (this.h - PaintPopup.hTab - 2 * AvMain.hDuBox - (wStr << 1))) < 0) {
         cmyLim = 0;
      }

      cmtoY = 0;
      cmy = 0;
      this.x = (Canvas.w - this.w) / 2;
      this.y = (Canvas.h - Canvas.hTab - this.h) / 2;
      this.time = System.currentTimeMillis();
   }

   public final void updateKey() {
      if (System.currentTimeMillis() - this.time >= 1000L) {
         this.time = System.currentTimeMillis();
      }

      ++this.count;
      boolean var1 = false;
      if (yL != 0) {
         yL += -yL >> 1;
      }

      if (yL == -1) {
         yL = 0;
      }

      if (Canvas.isPointerClick && Canvas.isPointer(this.x, this.y, this.w, this.h) && !this.trans) {
         this.pa = cmy;
         this.trans = true;
         this.vY = 0;
      }

      if (this.trans) {
         int var2 = Canvas.dy();
         if (Canvas.isPointerDown) {
            if (Canvas.gameTick % 3 == 0) {
               this.dyTran = Canvas.py;
               this.timePointY = this.count;
            }

            cmtoY = this.pa + var2;
            this.vY = 0;
            if (cmtoY < 0 || cmtoY > cmyLim) {
               cmtoY = this.pa + var2 / 2;
            }

            cmy = cmtoY;
         }

         if (Canvas.isPointerRelease) {
            this.trans = false;
            int var3 = this.count - this.timePointY;
            int var4;
            if (CRes.abs(var4 = this.dyTran - Canvas.py) > 40 && var3 < 10 && cmtoY > 0 && cmtoY < cmyLim) {
               this.vY = var4 / var3 * 10;
            }

            this.timePointY = -1;
            if (Math.abs(var2) < 10) {
               cmtoY = this.pa + var2;
            }
         }
      }

      if (Canvas.keyHold[2]) {
         cmtoY -= 14;
         var1 = true;
      } else if (Canvas.keyHold[8]) {
         var1 = true;
         cmtoY += 14;
      }

      if (var1) {
         if (cmtoY < 0) {
            cmtoY = 0;
         }

         if (cmtoY > cmyLim) {
            cmtoY = cmyLim;
         }
      }

      if (this.vY != 0) {
         if (cmy < 0 || cmy > cmyLim) {
            this.vY -= this.vY / 4;
            cmy += this.vY / 20;
            if (this.vY / 10 <= 1) {
               this.vY = 0;
            }
         }

         if (cmy < 0) {
            if (cmy < -this.h / 2) {
               cmy = -this.h / 2;
               cmtoY = 0;
               this.vY = 0;
            }
         } else if (cmy > cmyLim) {
            if (cmy < cmyLim + this.h / 2) {
               cmy = cmyLim + this.h / 2;
               cmtoY = cmyLim;
               this.vY = 0;
            }
         } else {
            cmy += this.vY / 10;
         }

         cmtoY = cmy;
         this.vY -= this.vY / 10;
         if (this.vY / 10 == 0) {
            this.vY = 0;
         }
      } else if (cmy < 0) {
         cmtoY = 0;
      } else if (cmy > cmyLim) {
         cmtoY = cmyLim;
      }

      if (cmy != cmtoY) {
         cmvy = cmtoY - cmy << 2;
         cmdy += cmvy;
         cmy += cmdy >> 4;
         cmdy &= 15;
      }

      super.updateKey();
   }

   public final void paint(Graphics var1) {
      Canvas.resetTrans(var1);
      Canvas.paint.paintBoxTab(var1, this.x, this.y, this.h, this.w, 0, 0, PaintPopup.gI().wSub, this.wTab, PaintPopup.hTab, 1, 1, PaintPopup.gI().count, PaintPopup.gI().colorTab, this.title);
      var1.setClip(this.x + 4, this.y + PaintPopup.hTab + 4 * AvMain.hd, this.w - 8, this.h - PaintPopup.hTab - 8 * AvMain.hd);
      var1.translate(this.x + this.xChange, this.y + PaintPopup.hTab);
      var1.translate(0, -cmy);
      var1.setColor(0);

      int var2;
      for(var2 = 0; var2 < this.listLabel.size(); ++var2) {
         StringObj var3;
         if ((var3 = (StringObj)this.listLabel.elementAt(var2)).y > cmy - 10 && var3.y < cmy + this.h) {
            int var4;
            if (var3.str.length() > 2 && var3.str.substring(0, 1).equals("¶")) {
               var4 = Integer.parseInt(var3.str.substring(1, var3.str.length()), 16);
               PaintPopup.fill(var3.x, var3.y, Canvas.w - (var3.x << 1), 1, var4, var1);
            } else {
               var4 = var3.x;
               if (var3.anthor == 2) {
                  var4 += (this.w - 30) / 2 + 4;
               } else if (var3.anthor == 1) {
                  var4 += this.w - 30 + 10;
               }

               if (var3.str.length() > 2 && var3.str.substring(0, 1).equals("Ę")) {
                  Canvas.fontChatB.drawString(var1, var3.str.substring(1, var3.str.length()), var4, var3.y, var3.anthor);
               } else if (var3.str.length() > 1 && var3.str.substring(0, 1).equals("0")) {
                  var1.setColor(8654855);
                  int var5 = Canvas.M.getWidth(var3.str.substring(1) + "") + 20;
                  var1.fillRect(var4 - var5 / 2, var3.y + AvMain.hNormal / 2 - AvMain.hSmall / 2 - 1, var5, Canvas.M.getHeight());
                  Canvas.M.drawString(var1, var3.str.substring(1) + "", var4, var3.y + AvMain.hNormal / 2 - AvMain.hBlack / 2, var3.anthor);
               } else {
                  Canvas.normalFont.drawString(var1, var3.str, var4, var3.y, var3.anthor);
               }
            }
         }
      }

      for(var2 = 0; var2 < this.listPic.size(); ++var2) {
         PictureObj var6;
         if ((var6 = (PictureObj)this.listPic.elementAt(var2)).b + var6.e > cmy && var6.b < cmy + this.h) {
            var1.drawImage((Image)this.listImg.get(String.valueOf(var6.d)), var6.a * ((this.w - (this.xChange << 1)) / 2), var6.b, var6.c);
         }
      }

      super.paint(var1);
   }
}

File: client/unity/Assets/Scripts/Assembly-CSharp/CustomTab.cs
using System;
using System.Globalization;

public class CustomTab : Face
{
	private class IActionCT : IAction
	{
		private readonly sbyte idAction;

		public IActionCT(sbyte idAction)
		{
			this.idAction = idAction;
		}

		public void perform()
		{
			GlobalService.gI().doCustomTab(idAction);
		}
	}

	public static CustomTab me;

	public int x;

	public int xChange;

	public int y;

	public int w;

	public int h;

	public MyVector listLabel = new MyVector();

	public MyVector listPic = new MyVector();

	public MyHashTable listImg;

	public string title = string.Empty;

	public string strTemp = string.Empty;

	public int selected;

	public int x0 = 5;

	public int y0 = 5;

	private int wTab = 30;

	public sbyte idAction;

	public sbyte countClose;

	private long time;

	public static int[] timeSub;

	public static int cmtoY;

	public static int cmy;

	public static int cmdy;

	public static int cmvy;

	public static int cmyLim;

	public static int yL;

	public static int wStr = 14;

	private int ww;

	private bool trans;

	private bool tranKey;

	private bool changeFocus;

	public int pa;

	public int pb;

	public int vY;

	public int dyTran;

	public bool transY;

	private long timePointY;

	private long count;

	public CustomTab()
	{
		setSize();
		wStr = Canvas.blackF.getHeight() + 5 * AvMain.hd;
	}

	public static CustomTab gI()
	{
		return (me != null) ? me : (me = new CustomTab());
	}

	private void setSize()
	{
		w = Canvas.w - 20;
		h = Canvas.hCan - 20;
	}

	public void init()
	{
		setSize();
		setInfo(listImg, title, strTemp, idAction);
	}

	private void setFont(string str)
	{
		int num = str.IndexOf("Ę");
		if (num != -1)
		{
			string str2 = str.Substring(0, num);
			setCanhle(str2, string.Empty);
			str = str.Substring(num + 1);
			int num2 = str.IndexOf("\n");
			if (num2 != -1)
			{
				string str3 = str.Substring(0, num2);
				setCanhle(str3, "Ę");
				str = str.Substring(num2 + 1);
				setFont(str);
			}
			else
			{
				setCanhle(str, "Ę");
			}
		}
		else
		{
			setCanhle(str, string.Empty);
		}
	}

	private void setCanhle(string str, string tem)
	{
		if (str.Equals(string.Empty))
		{
			return;
		}
		int num = str.IndexOf("ę");
		if (num != -1)
		{
			string text = str.Substring(0, num);
			if (!text.Equals(string.Empty))
			{
				addd(text, tem, 0);
			}
			int anthor = int.Parse(str.Substring(num + 1, 1));
			str = str.Substring(num + 2, str.Length - (num + 2));
			int num2 = str.IndexOf("\n");
			if (num2 != -1)
			{
				addd(str.Substring(0, num2), tem, anthor);
				setCanhle(str.Substring(num2 + 1), tem);
			}
			else
			{
				addd(str, tem, anthor);
			}
		}
		else
		{
			addd(str, tem, 0);
		}
	}

	private void addd(string str, string tem, int anthor)
	{
		int num = 0;
		string[] array = null;
		if (str.IndexOf("tem") != -1)
		{
			num = 1;
			array = Canvas.blackF.splitFontBStrInLine(str, w - 60 * AvMain.hd);
		}
		else
		{
			array = Canvas.tempFont.splitFontBStrInLine(str, w - 60 * AvMain.hd);
		}
		for (int i = 0; i < array.Length; i++)
		{
			StringObj stringObj = new StringObj(x0, y0 += wStr, tem + array[i]);
			stringObj.anthor = anthor;
			listLabel.addElement(stringObj);
		}
	}

	public void setInfo(MyHashTable list, string title, string str, sbyte idAction)
	{
		count = 0L;
		ww = 0;
		center = null;
		if (idAction != -1)
		{
			center = new Command(T.selectt, new IActionCT(idAction));
		}
		this.idAction = idAction;
		ww = 0;
		wTab = Canvas.tempFont.getWidth(title) + 80 * AvMain.hd;
		if (wTab < 80 * AvMain.hd)
		{
			wTab = 80 * AvMain.hd;
		}
		listImg = list;
		this.title = title;
		strTemp = str;
		listLabel.removeAllElements();
		listPic.removeAllElements();
		bool flag = false;
		x0 = 0;
		y0 = -10;
		while (true)
		{
			int num = str.IndexOf("µ");
			if (num == -1)
			{
				break;
			}
			string text = str.Substring(0, num);
			str = str.Substring(num + 1, str.Length - (num + 1));
			if (flag)
			{
				int num2 = text.IndexOf(",");
				string text2 = text.Substring(0, num2);
				text = text.Substring(num2 + 1, text.Length - (num2 + 1));
				int num3 = text.IndexOf(",");
				int num4 = int.Parse(text.Substring(0, num3));
				text = text.Substring(num3 + 1, text.Length - (num3 + 1));
				bool flag2 = int.Parse(text) != 0;
				Image image = (Image)listImg.get(text2 + string.Empty);
				int num5 = 0;
				switch (num4)
				{
				case 17:
					num5 = 1;
					break;
				case 24:
					num5 = 2;
					break;
				}
				PictureObj pictureObj = new PictureObj(int.Parse(text2), num5, y0 + wStr + 5, num4, flag2);
				pictureObj.w = image.getWidth();
				pictureObj.h = image.getHeight();
				if (flag2)
				{
					PictureObj pictureObj2 = (PictureObj)listPic.elementAt(listPic.size() - 1);
					int height = ((Image)listImg.get(pictureObj2.ID + string.Empty)).getHeight();
					if (image.getHeight() > height)
					{
						pictureObj2.y += image.getHeight() - height;
					}
					pictureObj.y = pictureObj2.y + height - image.getHeight();
				}
				y0 = pictureObj.y + image.getHeight() - 10;
				image = null;
				listPic.addElement(pictureObj);
				text = string.Empty;
			}
			flag = !flag;
			while (true)
			{
				int num6 = text.IndexOf("¶");
				if (num6 != -1)
				{
					string text3 = text.Substring(0, num6);
					text = text.Substring(num6 + 1, text.Length - (num6 + 1));
					try
					{
						setFont("¶" + text3);
						y0 -= wStr / 2;
					}
					catch (Exception e)
					{
						Out.logError(e);
						setFont(text3);
					}
				}
				else
				{
					if (!text.Equals(string.Empty))
					{
						setFont(text.Substring(0, text.Length - 1));
					}
					if (str.IndexOf("µ") != -1 || str.IndexOf("¶") == -1)
					{
						break;
					}
					text = str;
					str = string.Empty;
				}
			}
		}
		setFont(str);
		xChange = 9 * AvMain.hd;
		int num7 = 0;
		for (int i = 0; i < listLabel.size(); i++)
		{
			StringObj stringObj = (StringObj)listLabel.elementAt(i);
			FontX fontX = null;
			fontX = ((stringObj.str.IndexOf("tem") == -1) ? Canvas.tempFont : Canvas.blackF);
			int width = fontX.getWidth(stringObj.str);
			if (width + 100 * AvMain.hd > num7)
			{
				num7 = width + 100 * AvMain.hd;
			}
		}
		if (num7 < w)
		{
			w = num7;
			xChange = 10 * AvMain.hd;
		}
		if (y0 + 10 + wStr * 2 < h - 30)
		{
			h = y0 + 10 + wStr * 2 + 20;
		}
		if (w > Canvas.w - 100 * AvMain.hd)
		{
			w = Canvas.w;
			h = Canvas.hCan;
		}
		if (w < wTab * 2)
		{
			w = wTab * 2;
		}
		if (w > Canvas.w)
		{
			w = Canvas.w;
		}
		cmyLim = y0 - (h - PaintPopup.hTab - 2 * AvMain.hDuBox - wStr * 2);
		if (cmyLim < 0)
		{
			cmyLim = 0;
		}
		cmy = (cmtoY = 0);
		x = (Canvas.w - w) / 2;
		y = (Canvas.hCan - h) / 2;
		time = Environment.TickCount;
	}

	public override void updateKey()
	{
		if (Environment.TickCount - time >= 1000)
		{
			if (timeSub != null)
			{
				for (int i = 0; i < timeSub.Length; i++)
				{
					if (timeSub[i] > 0)
					{
						timeSub[i]--;
						if (timeSub[i] == 0)
						{
							FarmService.gI().doStealInfo();
						}
					}
				}
			}
			time = Environment.TickCount;
		}
		count++;
		bool flag = false;
		if (yL != 0)
		{
			yL += -yL >> 1;
		}
		if (yL == -1)
		{
			yL = 0;
		}
		int num = Canvas.w - 45 * AvMain.hd;
		int num2 = 15 * AvMain.hd;
		if (w != Canvas.w)
		{
			num = x + w - 20 * AvMain.hd;
			num2 = y + 20 * AvMain.hd;
		}
		if (Canvas.isPointerClick && Canvas.isPointer(num, num2, 40 * AvMain.hd, 40 * AvMain.hd))
		{
			countClose = 5;
			tranKey = true;
			Canvas.isPointerClick = false;
		}
		if (tranKey)
		{
			if (Canvas.isPointerDown && !Canvas.isPointer(num, num2, 40 * AvMain.hd, 40 * AvMain.hd))
			{
				countClose = 0;
			}
			if (Canvas.isPointerRelease)
			{
				Canvas.isPointerRelease = false;
				tranKey = false;
				if (countClose == 5)
				{
					countClose = 0;
					listLabel.removeAllElements();
					listPic.removeAllElements();
					listImg.clear();
					close();
					Canvas.endDlg();
					me = null;
				}
			}
		}
		if (Canvas.isPointerClick && Canvas.isPointer(x, y, w, h) && !trans)
		{
			pa = cmy;
			trans = true;
			vY = 0;
		}
		if (trans)
		{
			if (Canvas.isPointerDown)
			{
				int num3 = Canvas.dy();
				if (Canvas.gameTick % 3 == 0)
				{
					dyTran = Canvas.py;
					timePointY = count;
				}
				vY = 0;
				cmtoY = pa + num3;
				if (cmtoY < 0 || cmtoY > cmyLim)
				{
					cmtoY = pa + num3 / 2;
				}
				cmy = cmtoY;
			}
			if (Canvas.isPointerRelease)
			{
				int num4 = (int)(count - timePointY);
				int num5 = dyTran - Canvas.py;
				if (CRes.abs(num5) > 40 && num4 < 10 && cmtoY > 0 && cmtoY < cmyLim)
				{
					vY = num5 / num4 * 10;
				}
				timePointY = -1L;
				trans = false;
			}
		}
		if (Canvas.keyHold[2])
		{
			cmtoY -= 14;
			flag = true;
		}
		else if (Canvas.keyHold[8])
		{
			flag = true;
			cmtoY += 14;
		}
		if (flag)
		{
			if (cmtoY < 0)
			{
				cmtoY = 0;
			}
			if (cmtoY > cmyLim)
			{
				cmtoY = cmyLim;
			}
		}
		if (cmy != cmtoY)
		{
			cmvy = cmtoY - cmy << 2;
			cmdy += cmvy;
			cmy += cmdy >> 4;
			cmdy &= 15;
		}
		if (Math.abs(cmtoY - cmy) < 15 && cmy < 0)
		{
			cmtoY = 0;
		}
		if (Math.abs(cmtoY - cmy) < 10 && cmy > cmyLim)
		{
			cmtoY = cmyLim;
		}
		if (vY != 0)
		{
			int num6 = h - PaintPopup.hTab - 8 * AvMain.hd;
			if (cmy < 0 || cmy > cmyLim)
			{
				vY -= vY / 4;
				cmy += vY / 20;
				if (vY / 10 <= 1)
				{
					vY = 0;
				}
			}
			if (cmy < 0)
			{
				if (cmy < -num6 / 2)
				{
					cmy = -num6 / 2;
					cmtoY = 0;
					vY = 0;
				}
			}
			else if (cmy > cmyLim)
			{
				if (cmy < cmyLim + num6 / 2)
				{
					cmy = cmyLim + num6 / 2;
					cmtoY = cmyLim;
					vY = 0;
				}
			}
			else
			{
				cmy += vY / 10;
			}
			cmtoY = cmy;
			vY -= vY / 10;
			if (vY / 10 == 0)
			{
				vY = 0;
			}
		}
		else if (cmy < 0)
		{
			cmtoY = 0;
		}
		else if (cmy > cmyLim)
		{
			cmtoY = cmyLim;
		}
		base.updateKey();
	}

	public override void paint(MyGraphics g)
	{
		Canvas.resetTransNotZoom(g);
		Canvas.paint.paintTransBack(g);
		Canvas.paint.paintBoxTab(g, x, y, h, w, 0, PaintPopup.gI().wSub, wTab, 40 * AvMain.hd, 1, 1, PaintPopup.gI().count, PaintPopup.gI().colorTab, title, -1, -1, false, (w == Canvas.w) ? true : false, null, 0f, null);
		g.setClip(x + 4, y + 45 * AvMain.hd, w - 8, h - 40 * AvMain.hd - 10);
		g.translate(x + xChange, y + 35 * AvMain.hd);
		g.translate(0f, -cmy);
		g.setColor(0);
		int num = 0;
		for (int i = 0; i < listLabel.size(); i++)
		{
			StringObj stringObj = (StringObj)listLabel.elementAt(i);
			if (stringObj.y <= cmy - 10 || stringObj.y >= cmy + h)
			{
				continue;
			}
			if (stringObj.str.Length > 2 && stringObj.str.Substring(0, 1).Equals("¶"))
			{
				int color = int.Parse(stringObj.str.Substring(1, stringObj.str.Length - 1), NumberStyles.HexNumber);
				PaintPopup.fill(stringObj.x, stringObj.y, Canvas.w - stringObj.x * 2, 1, color, g);
				continue;
			}
			int num2 = stringObj.x;
			if (stringObj.anthor == 2)
			{
				num2 += (w - 30) / 2 + 4;
			}
			else if (stringObj.anthor == 1)
			{
				num2 += w - 30 + 10;
			}
			if (stringObj.str.Length > 2 && stringObj.str.Substring(0, 1).Equals("Ę"))
			{
				Canvas.blackF.drawString(g, stringObj.str.Substring(1, stringObj.str.Length - 1), num2, stringObj.y, stringObj.anthor);
			}
			else if (stringObj.str.Length > 1 && stringObj.str.Substring(0, 1).Equals("0"))
			{
				Canvas.smallFontYellow.drawString(g, stringObj.str.Substring(1) + ((timeSub == null || timeSub[num] < 0) ? string.Empty : (" " + timeSub[num] / 60 + ":" + (timeSub[num] - timeSub[num] / 60 * 60))), num2, stringObj.y + AvMain.hNormal / 2 - AvMain.hSmall / 2 + 4, stringObj.anthor);
				num++;
			}
			else
			{
				Canvas.tempFont.drawString(g, stringObj.str, num2, stringObj.y, stringObj.anthor);
			}
		}
		for (int j = 0; j < listPic.size(); j++)
		{
			PictureObj pictureObj = (PictureObj)listPic.elementAt(j);
			if (pictureObj.y + pictureObj.h > cmy && pictureObj.y < cmy + h)
			{
				g.drawImage((Image)listImg.get(pictureObj.ID + string.Empty), pictureObj.x * ((w - xChange * 2) / 2), pictureObj.y, pictureObj.orthor);
			}
		}
		base.paint(g);
		if ((w == Canvas.w) ? true : false)
		{
			g.drawImage(ListScr.imgCloseTabFull[countClose / 3], Canvas.w - 25 * AvMain.hd, 35 * AvMain.hd, 3);
		}
		else
		{
			g.drawImage(ListScr.imgCloseTab[countClose / 3], x + w - 2 * AvMain.hd, y + 40 * AvMain.hd, 3);
		}
	}
}
